Remove roaring.Bitmaps from all of the authz/permissions code

varsanojidan opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/33172

I did not touch any roaring.Bitmap code related to zoektquery because that library/struct requires roaring.Bitmaps.

## Test plan
Verified all existing testing still passes after removal of bitmaps.

varsanojidan commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -20,7 +19,7 @@ var _ graphqlbackend.RepositoryConnectionResolver = &repositoryConnectionResolve
 type repositoryConnectionResolver struct {
 	db  database.DB
-	ids *roaring.Bitmap
+	ids []int32 //sorted slice of IDs
```
On thing id like some input on is this right here, this used to be a roaring bitmap that was converted into an array of ints (guaranteed to be sorted), and I modified the field to an array of int32s and I think I guarantee its sorted through the usage of my helper method:  GenerateSortedIDsSlice() []int32 throughout the code.

I could guarantee its in sorted order by just sorting it first thing in the compute() method in this same file, but its probably redundant, so im curious if anyone has any opinions here.

ryanslade replied:
I think the way you've done it looks good 👍

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -37,37 +36,39 @@ type repositoryConnectionResolver struct {
+		idsSize := int32(len(r.ids))
+
+		// Generate a slice of repo IDs ranging from index after+1 to: after+first or until the end of the slice, whichever is less.
+		if idsSize >= int32(afterID)+1 {
```
This logic seems to be taken from the old code but is very confusing to me. From what I can tell, `afterID` is a repo id from the database, which happens to be a 32bit int.

We're then adding 1 to that id? That seems to assume that there are no gaps in our id table which isn't true since repos can be deleted.

We're also treating `afterID` as an index into a slice on line 52.

I have a feeling this code might pass tests where we're creating a few rows with no gaps but in production it probably won't work.

varsanojidan replied:
Yeah I completely agree with you, when I was reading the original code I had a hard time discerning what exactly after/afterID were supposed to represent because of the same reason you listed, the adding 1 to it didnt make sense if it wasnt an index because like you I was wondering what if that repoID didnt exist, which is why I was treating it like an index here. But looking back at the original code, you're absolutely right its supposed to represent the ID not the index, which is super easy to do in the sorted slice that we have now, but very confusing at how exactly it functioned previously

unknwon replied:
`AdvanceIfNeeded` advances based on value not on index.

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -37,37 +36,39 @@ type repositoryConnectionResolver struct {
 
-		if iter.HasNext() {
+		// No more repo IDs to paginate through.
+		if idsSize <= int32(afterID)+r.first {
```
Similar to the previous comment, it looks like we're treating afterID as an slice index here rather than a database id.

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/users.go:
```diff
@@ -37,22 +36,26 @@ type userConnectionResolver struct {
-			userIDs = append(userIDs, int32(iter.Next()))
-			if len(userIDs) >= int(r.first) {
-				break
+		idsSize := int32(len(r.ids))
```
Same as other comments above

ryanslade reviewed (changes requested):
Looks good apart from my comments about the pagination stuff in the resolver.

ryanslade commented:
@varsanojidan Please let me know when you're ready for another review 🔍

varsanojidan commented:
> @varsanojidan Please let me know when you're ready for another review 🔍

Ready for another review :)

ryanslade commented on enterprise/internal/database/perms_store.go:
```diff
@@ -4,6 +4,7 @@ import (
+	"sort"
 	"time"
 
 	"github.com/RoaringBitmap/roaring"
```
It looks like we only use this in a couple more places in this file and it can be refactored away.

ryanslade replied:
Can't comment as it's not in this diff, but the command on load() in this file should also be updated to remove the mention of roaring bitmaps.

varsanojidan replied:
whoops, cant believe i missed that lol, will fix!

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -37,37 +36,55 @@ type repositoryConnectionResolver struct {
 	r.once.Do(func() {
-		// Create the bitmap iterator and advance to the next value of r.after.
 		var afterID api.RepoID
+		afterIDIdx := 0
```
Still reading through this and I think I understand why I'm finding it so confusing.

The way this resolver works is that the `ids` slice will contain all ids and we then only resolve a subset of those ids depending on the values in `first` and `after`. I think it could be really helpful to add a comment about this above the definition of the `ids` and `after` fields as it's not obvious.

ryanslade commented:
Could you add some tests in `resolve_test.go`? It doesn't look like we have anything that checks the pagination is working the way we expect. I think we might be able to simplify this code at some point and it would be good to verify that it's working before we do.

varsanojidan commented:
@ryanslade addressed your comments and added some tests, lmk what you think

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -35,39 +35,65 @@ type repositoryConnectionResolver struct {
 // is the site admin because this method computes data from all available information in
 // the database.
+// This function takes returns a pagination of the repo IDs
+// r.ids - the full slice of sorted repo IDs
+// r.after - (optional) the repo ID to start the paging after (does not include the after ID itself)
+// r.first - the # of repo IDs to return
```
FYI this would not result in a good-looking formatting in almost all code-intel features, e.g. in Sourcergaph

<img width="610" alt="CleanShot 2022-04-06 at 11 08 42@2x" src="https://user-images.githubusercontent.com/2946214/161888407-a65133ae-d1bd-4b05-8793-7ef784f0ead4.png">

An indentation is needed for preserving the line separators:

```suggestion
//  r.ids - the full slice of sorted repo IDs
//  r.after - (optional) the repo ID to start the paging after (does not include the after ID itself)
//  r.first - the # of repo IDs to return
```

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -35,39 +35,65 @@ type repositoryConnectionResolver struct {
-			afterID, r.err = graphqlbackend.UnmarshalRepositoryID(graphql.ID(*r.after))
-			if r.err != nil {
+		var idSubset []int32
+		found := false
```
nit: `found` is only used in the `else`-branch, so we can move this down.

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -35,39 +35,65 @@ type repositoryConnectionResolver struct {
-			repoIDs = append(repoIDs, api.RepoID(iter.Next()))
-			if len(repoIDs) >= int(r.first) {
-				break
+			for idx, id := range r.ids {
+				if id == int32(afterID) {
+					found = true
```
From the client's POV, this could break previously working query due to legitimated repo deletions.

For example, if I am querying to list first 2 repos after ID 10, semantically the API should give me 2 repos as long as their IDs are greater than 10, regardless of whether 10 actually exists, e.g. [11, 12], or [13, 18].

Here is what I suggest:
```suggestion
			for idx, id := range r.ids {
				if id >= int32(afterID) {
					found = true
```

varsanojidan replied:
Great call on the change in behavior, I addressed the comments you made in my most recent commit, lmk what you think

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -2,15 +2,14 @@ package resolvers
 
 import (
 	"context"
+	"github.com/sourcegraph/sourcegraph/internal/api"
 	"sync"
 
-	"github.com/RoaringBitmap/roaring"
 	"github.com/graph-gophers/graphql-go"
 
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/backend"
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend"
```
nit: import grouping, the added line should be in the third group since it is not a stdlib package.

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -35,39 +34,62 @@ type repositoryConnectionResolver struct {
+			repoIDs[i] = api.RepoID(idSubset[i])
+		}
 
-		// TODO(asdine): GetByIDs now returns the complete repo information rather that only a subset.
-		// Ensure this doesn't have an impact on performance and switch to using ListMinimalRepos if needed.
```
qq: Is this comment no longer a concern?

varsanojidan replied:
This was an accidental removal, ill add it back in

unknwon reviewed (commented):
Thanks for the followup!

Another general question, we're already using int64 for pending permissions in DB, are using `int32` in application safe? I don't think so... previously roaring bitmap uses uint32 so they're at least doubled range of int32 (though still smaller than int64 and can cause int overflow some point in the future)

varsanojidan commented:
@unknwon regarding your concerns about int32 vs uint32, I believe the whole reason uint32 was used at all (like you mentioned) was just because of roaring.Bitmaps returning uint32s. The reason I felt int32 would be safe was because from what I saw, the userIDs and repoIDs were cast back to or from int32s to begin with like here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L682&subtree=true

and here (api.RepoID also being int32):
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L440&subtree=true

From what I can tell UserID/RepoID always start/end as int32s

unknwon commented:
> @unknwon regarding your concerns about int32 vs uint32, I believe the whole reason uint32 was used at all (like you mentioned) was just because of roaring.Bitmaps returning uint32s. The reason I felt int32 would be safe was because from what I saw, the userIDs and repoIDs were cast back to or from int32s to begin with like here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L682&subtree=true
> 
> and here (api.RepoID also being int32): https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/repo-updater/internal/authz/perms_syncer.go?L440&subtree=true
> 
> In the second example int32 repo IDs are converted to uint32s just for the sake of the methods return.
> 
> I assumed because the IDs originate from or are converted to int32s in a large amount of places in the code that it would be safe, though I do agree that it's kind of not great to be using types so haphazardly like this, though the way to fix it would be to scour the code base and change all the types to either int32 or uint32, might be worth to make a tech debt ticket for it to minimize the impact of this PR.
> 
> What do you think?

For non-pending permissions I totally agree with you. What about pending user IDs? IIRC we have started storing them as int64 in database, so I think casting them to int32 would cause overflow at least for the customer that int64 patch was for.

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -35,25 +34,49 @@ type repositoryConnectionResolver struct {
 // 🚨 SECURITY: It is the caller's responsibility to ensure the current authenticated user
 // is the site admin because this method computes data from all available information in
 // the database.
+// This function takes returns a pagination of the repo IDs
```
```suggestion
// This function returns a pagination of the repo IDs
```

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/users.go:
```diff
@@ -35,39 +33,59 @@ type userConnectionResolver struct {
 // 🚨 SECURITY: It is the caller's responsibility to ensure the current authenticated user
 // is the site admin because this method computes data from all available information in
 // the database.
+// This function takes returns a pagination of the user IDs
```
```suggestion
// This function returns a pagination of the user IDs
```

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/users.go:
```diff
@@ -35,39 +33,59 @@ type userConnectionResolver struct {
 // is the site admin because this method computes data from all available information in
 // the database.
+// This function takes returns a pagination of the user IDs
+// 	r.ids - the full slice of sorted user IDs
```
nittiest of nits: The spacing here is different to the similar comment above.

varsanojidan replied:
Is the issue you're referring to the indenting of the lines in the comments? It was changed with this suggestion: https://github.com/sourcegraph/sourcegraph/pull/33333#discussion_r843420828

ryanslade commented:
I know it's a pain but I think we need to be really careful about overflows.

I would suggest that we don't try and group all of our helpers code into int32 versions and we should instead make int64 versions too. Ideally we don't want any code to be converting between int32 and int64 in either direction.

This will lead to a bit of code duplication but at least it's guaranteed to be safe. 

We can switch it all out to generic helpers soon anyway 😄

varsanojidan commented:
@unknwon oh I see, yeah youre completely right, the IDs for the resolvers could also be coming from the pending permissions table in which case they would be int64 and could have overflow issues, im going to fix up this code to address this, great catch!

@ryanslade I agree, not having generics here has been the absolute worst lol. Alright Ill do some refactoring of the code to avoid these conversions entirely :)

varsanojidan commented:
Hey @unknwon @ryanslade I've refactored perm_store.go toresolve the issue of casting between int64/int32/uint32, everything is int32.

- I removed the generic load function that was loading everything up as int64s and replaced it with 4 load functions that are tailored to the individual tables (user_permissions/repo_permissions/user_pending_permissions/repo_pending_permissions). With this change were able to pass back the actual data type as reflected in the DB and not have to worry about conversion.
- I removed uint32s because they were only really there for ease of use with roaring.Bitmaps and I changed them to int32, which is the actual type for those fields in the database

With this changes the code looks a lot better, no more casting all over the place and objects accurately represent whats in the db. Lmk what you guys think :)

unknwon commented on enterprise/internal/database/perms_store_test.go:
```diff
@@ -41,23 +41,22 @@ func cleanupPermsTables(t *testing.T, s *permsStore) {
-	bm := roaring.NewBitmap()
-	bm.AddMany(ids)
-	return bm
+func toMapSet(ids ...int32) map[int32]struct{} {
```
The case here seems being inconsistent with `mapsetToArray`
```suggestion
func toMapset(ids ...int32) map[int32]struct{} {
```

Or vice versa.

unknwon commented on enterprise/internal/database/perms_store_test.go:
```diff
@@ -280,11 +279,13 @@ func checkRegularPermsTable(s *permsStore, sql string, expects map[int32][]uint3
+		}
+
+		if cmp.Diff(expects[id], intIDs, cmpopts.SortSlices(comparator)) != "" {
+			return errors.Errorf("intIDs - key %v: want %q but got %q", id, expects[id], intIDs)
```
IIRC `%q` only works for strings, which were, but we have `[]uint32` now.
```suggestion
			return errors.Errorf("intIDs - key %d: want %d but got %v", id, expects[id], intIDs)
```

unknwon reviewed (approved):
Overall LGTM, nice effort!

ryanslade commented on enterprise/cmd/frontend/internal/authz/resolvers/repositories.go:
```diff
@@ -63,11 +86,12 @@ func (r *repositoryConnectionResolver) compute(ctx context.Context) ([]*types.Re
 
-		if iter.HasNext() {
+		// The last id in this page is the last id in r.ids, no more pages
+		if int32(repoIDs[len(repoIDs)-1]) == r.ids[len(r.ids)-1] {
```
I guess with generics we'll be able to have a helper function like `slice.Last`

ryanslade commented on enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:
```diff
@@ -625,13 +624,15 @@ func (s *PermsSyncer) syncUserPerms(ctx context.Context, userID int32, noPerms b
+	externalRepoIds := append(externalAccountsRepoIDs, externalServicesRepoIDs...)
+	for i := range externalRepoIds {
+		p.IDs[int32(externalRepoIds[i])] = struct{}{}
 	}
-	p.IDs.AddMany(externalAccountsRepoIDs)
-	p.IDs.AddMany(externalServicesRepoIDs)
```
Ah, so you mean loop over `externalAccountsRepoIDs ` and `externalServicesRepoIDs` adding to the map?

Yeah, that probably makes sense. No need for the temporarily used `externalRepoIds`

unknwon replied:
> Ah, so you mean loop over `externalAccountsRepoIDs ` and `externalServicesRepoIDs` adding to the map?

Yes, since I'd expect the list of IDs could be large.

ryanslade reviewed (approved):
Nice!

And thanks to @unknwon for such a thorough review I don't have much to add :)

unknwon commented on enterprise/cmd/frontend/internal/authz/resolvers/resolver_test.go:
```diff
@@ -921,3 +1095,7 @@ func TestResolver_SetSubRepositoryPermissionsForUsers(t *testing.T) {
 	})
 }
+
+func convertToBase64(input string) string {
+	return base64.StdEncoding.EncodeToString([]byte(input))
+}
```
We have helpers likes `graphqlbackend.MarshalUserID` available 😁

unknwon commented on enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:
```diff
@@ -625,22 +624,25 @@ func (s *PermsSyncer) syncUserPerms(ctx context.Context, userID int32, noPerms b
-		p.IDs.Add(uint32(repoIDs[i]))
+		p.IDs[int32(repoIDs[i])] = struct{}{}
+	}
+	for i := range externalAccountsRepoIDs {
+		p.IDs[int32(externalAccountsRepoIDs[i])] = struct{}{}
+	}
+	for i := range externalServicesRepoIDs {
+		p.IDs[int32(externalServicesRepoIDs[i])] = struct{}{}
```
nit: a code comment explaining why we choose to use two loops.

unknwon reviewed (approved):
Did not do as through look as last review but I trust you 😁
